plugins/rknn/src/rknn/plugin.py

```diff
@@ -0,0 +1,119 @@
+import asyncio
+import concurrent.futures
+import os
+import platform
+import queue
+import threading
+from typing import Any, Coroutine, List, Tuple
+import urllib.request
+
+import numpy as np
+from PIL.Image import Image
+from rknnlite.api import RKNNLite
+
+from predict import PredictPlugin, Prediction
+from predict.rectangle import Rectangle
+
+# for Rockchip-optimized models, the postprocessing is slightly different from the original models
+from .optimized.yolo import post_process, IMG_SIZE, CLASSES
+
+
+rknn_verbose = False
+lib_download = 'https://github.com/airockchip/rknn-toolkit2/raw/v2.0.0-beta0/rknpu2/runtime/Linux/librknn_api/aarch64/librknnrt.so'
+model_download = 'https://github.com/bjia56/scrypted-rknn/raw/main/models/yolov6n_rk3588_optimized.rknn'
+lib_path = '/usr/lib/librknnrt.so'
+
+
+def ensure_compatibility():
+    err_msg = 'RKNN plugin is only supported on Linux/ARM64 platform with RK3588(S) CPU'
+    if platform.machine() != 'aarch64':
+        raise RuntimeError(err_msg)
+
+    if platform.system() != 'Linux':
+        raise RuntimeError(err_msg)
+
+    try:
+        with open('/proc/device-tree/compatible') as f:
+            if not 'rk3588' in f.read():
+                raise RuntimeError(err_msg)
+    except IOError as e:
+        print('Failed to read /proc/device-tree/compatible: {}'.format(e))
+        print('If you are running this on RK3588(S) via Docker, ensure you are launching the container with --privileged option')
+        raise
+
+
+def thread_executor_main(model_path, core_num, core_mask, job_queue: queue.Queue):
+    rknn = RKNNLite(verbose=rknn_verbose)
+    ret = rknn.load_rknn(model_path)
+    if ret != 0:
+        raise RuntimeError('Failed to load model: {}'.format(ret))
+
+    ret = rknn.init_runtime(core_mask=core_mask)
+    if ret != 0:
+        raise RuntimeError('Failed to init runtime: {}'.format(ret))
+
+    print('RKNNLite runtime initialized on core {}'.format(core_num))
+
+    while True:
+        input_tensor, done_future = job_queue.get()
+
+        # check for shutdown
+        if input_tensor is None:
+            break
+
+        outputs = rknn.inference(inputs=[input_tensor])
+        done_future.set_result(outputs)
+
+
+class RKNNPlugin(PredictPlugin):
+    labels = {i: CLASSES[i] for i in range(len(CLASSES))}
+    jobs: queue.Queue
+    executors: List[threading.Thread]
+
+    def __init__(self, nativeId=None):
+        super().__init__(nativeId)
+        ensure_compatibility()
+
+        if not os.path.exists(lib_path):
+            print('Downloading librknnrt.so from {}'.format(lib_download))
+            urllib.request.urlretrieve(lib_download, lib_path)
```

Comment:
us downloadFile? it handles versioning and caching.

Reply:
It seems that the rknn runtime library requires this to be in /usr/lib, I assume they are doing a dynamic runtime load with a full path using dlopen, since changing LD_LIBRARY_PATH to include the downloadFile directory does not work. Auditwheel (standard python tool to copy shared objects into the wheel) also ignores this library, so it is likely loaded as a runtime dependency